llext-edk: rework, read data from `build_info.yml` and `.config`

pillo79 opened the pull request:
This PR improves the LLEXT EDK in several ways:
* an internal rework to reduce code duplication and improve extensibility
* read existing data from the current build artifacts, instead of using arguments

It is effectively a subset of #78727 that does not add any new information to the LLEXT EDK.

pillo79 committed:
llext-edk: refact: move file generation into functions

This commit refactors the code that generates the Makefile and CMake
EDK files into functions that accept the file type as an argument.

No functional changes are introduced in this commit.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext-edk: import data from build_info.yml and .config

Currently, the llext-edk.cmake script requires a number of variables to
be passed in from the main CMakeLists.txt file as arguments to be able
to customize the generated files.

To improve this rigid approach, the script is modified to read in the
following files in the build directory:

 * 'zephyr/.config', for the final set of Kconfig options used;
 * 'build_info.yml', for the cmake-related variables.

This is more flexible and also easier to maintain, as it doesn't require
manual changes to the main CMakelists.txt file when new variables need
to be referenced.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
cmake: yaml: use APPEND PROPERTY to shorten code

Use a more concise way to append to a list property, as suggested in a
previous PR review.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Review ready. Rebased and included minor fix from [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/80007#discussion_r1935441072) in related PR.

pillo79 commented:
@tejlmand please review. This is the last piece of the rework to use YAML in the EDK, no functional changes.

tejlmand reviewed (changes requested):
Generally looks good, though care must be taken wrt. use of `cmake_parse_arguments()`.
See relevant comments.

I also wonder if it is important in this case to have `GENEX` as a flag, or if we should simplify so that we just have:
```
#   build_info(<tag>... VALUE <value>... )
#   build_info(<tag>... GENEX <genex>... )
#   build_info(<tag>... PATH  <path>... )
```
that way the first keyword is still just separating `<tag>` from `<values>`.


There is also a possibility to detect genex'es, like this:
```
string(GENEX_STRIP "${value}" value_genex_free)
if(NOT value STREQUAL value_genex_free
  # Contain genex'es
endif()
```
which of course cost a little, but allows the caller of `build_info()` to just use or not-use genex'es as they like, just like they can do with other native CMake functions.

pillo79 committed:
cmake: support generator expressions in build_info()

This patch enables support for genexes in the 'build_info()' function by
identifying when arguments contain them and passing the GENEX flag to
'yaml_set()'.

Genexes are not supported in tag lists; using them will immediately
trigger an error by the schema checks.

They are also not currently supported with PATH entries, because the
actual contents of the list (whose paths are to be converted) are not
known until after the CMake code has been processed.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext-edk: refact: do not mix code and function definitions

Prepare the current code so that functions come before actual code.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 committed:
llext-edk: merge target-specific file generation

This commit removes all CMake- or Makefile-specific variables from the
code and instead uses a single list for each type of flag. Where a
difference is needed between the two, a special marker is used.
These markers are replaced later in the target-specific area of the
write functions.

Escaping of problematic characters is also added to the write functions,
to ensure that quotes and backslashes in the flags are properly handled.

Signed-off-by: Luca Burelli <l.burelli@arduino.cc>

pillo79 commented:
Thanks @tejlmand for the review, I did not consider the pitfalls you mentioned with `cmake_parse_arguments`.

Here is **v2**, addressing all concerns:
* use the current ad-hoc parser in `build_info()`
* automatically add the `GENEX` flag when required
* moved function definitions in `llext-edk.cmake` to the beginning of the file, as most were added in this PR

tejlmand reviewed (changes requested):
Looks good, though I suspect `GENEX` is not supposed to be part of the value :wink:

pillo79 commented:
Lunch coffee lost its effect by 7PM :man_facepalming: fixed, thanks for spotting!

teburd commented on cmake/llext-edk.cmake:
```diff
@@ -81,6 +81,17 @@ function(relative_dir dir relative_out bindir_out)
+#
+# Escape problematic characters in the string <str_in> and store the result in
+# <str_out>. The escaping is done to make the string suitable for <target>.
+function(edk_escape target str_in str_out)
```
something here is fundamentally wrong, why are we escaping escapes? Is this needed because of cmake? shell?

pillo79 replied:
:grinning: Yes!
A CMake string can contain any character (`\` **in a variable** is a literal backslash, it only needs two **in the source file**), but the file I am writing will be read by either Makefile or CMake, and both of them will swallow both `\` and `"` unless escaped.

teburd replied:
Somedays I question if cmake is actually beneficial in any form. Truly.

teburd commented on cmake/modules/extensions.cmake:
```diff
@@ -3737,6 +3737,14 @@ function(build_info)
     message(FATAL_ERROR "${CMAKE_CURRENT_FUNCTION}(...) missing a required argument: VALUE or PATH")
   endif()
 
+  string(GENEX_STRIP "${arg_list}" arg_list_no_genexes)
```
Fun and leads to this confusing set of changes and branches, but it is what it is I guess
